keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,566 @@
+have hit the parallelism limit. Once another Pod finishes, the Job controller
+syncs the Job, scanning for unsatisfied indexes and creates the missing Pods for
+them.
+
+The kubelet handles container restarts as usual, according to the
+`spec.template.spec.restartPolicy`.
+
+<<[UNRESOLVED TBD Beta: Track completed indexes in Job status]>>
+Once [kubernetes/kubernetes#28486](https://github.com/kubernetes/kubernetes/issues/28486)
+is resolved:
+
+The Job controller keeps track of completed indexes in the Job status in a
+base64 encoded bitmask.
+The command `kubectl describe` displays the list of completed indexes in a
+human-readable compressed format. For example:
+
+```
+Completed Indexes: [1-25,28,30-32]
+``` 
+
+The command crops the list of indexes once it reaches 200 characters.
+
+```
+Completed Indexes: [1-25,28,30-32,...]
+``` 
+<<[/UNRESOLVED]>>
+
+### Job parallelism
+
+A user can change the number of active Pods for a Job changing `.spec.parallelism`
+(note that `.spec.completions` is an immutable field).
+
+When starting a Job or increasing the parallelism, the Job controller creates
+Pods with lower completion index first, as long as there is no other completed
+or running Pod with the same index. This is to make the controller behavior more
+predictable. We do not offer guarantees on creation order based on completion
+index.
+
+Scale down is unaffected by completion index.
+
+### Test Plan
+
+Unit, integration and E2E tests cover the following Indexed Job mechanics:
+
+  - Creation with indexed Pod names and index annotations.
+  - Scale up and down.
+  - Pod failures.
+  
+Additionally, we add unit tests for API defaulting and validation, with feature
+gate enabled and disabled.
+  
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Complete features:
+  - Completion index in annotation
+  - Restart policy
+- Gather feedback from end users and operators' developers. Open questions:
+  - Are stable Pod names necessary?
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance
+- 2 examples of operators using Indexed Jobs.
+
+### Upgrade / Downgrade Strategy
+
+In the event of a kube-controller-manager upgrade, there should not be any
+existing Indexed Jobs with running Pods.
+
+In the event of a downgrade, existing Indexed Jobs will run as NonIndexed Jobs.
+The controller will track the existing Pods ignoring the completion index. New
+Pods will be created without a completion index. Existing workloads that
+expected the completion index will fail. But this is expected in a downgrade.
+
+In the event of an upgrade after a downgrade, the controller will ignore
+existing Pods without completion index for existing Indexed Jobs. This might
+cause a load spike in the cluster.
+
+If, instead of a downgrade, the feature is disabled:
+
+  - kube-apiserver sets `.spec.completionMode=NonIndexed` for new Jobs.
```

Comment:
I don't understand how this is done (and how you distinguish downgrade from disabling exactly). Can you be more specific?

Reply:
This is standard logic in API strategies for disabled feature gates. Reworded a bit.

I don't exactly distinguish a downgrade, as I cannot control the code of version N-1. So the paragraphs above describe how the code behaves in that scenario.

But what I can control is a feature being disabled, in which case I prefer to ignore Indexed Jobs that were already created.